Close the socket if there's a failure in start_connection()

top-oai opened the pull request:
<!-- Thank you for your contribution! -->

## What do these changes do?

Over in https://github.com/MagicStack/uvloop/issues/653 in issue is described where after upgrading from aiohttp 3.9.5 -> 3.10.5 (but same thing on 3.11.11), under certain conditions where exceptions are raised at the right/wrong time, continued use of `aiohttp` can result in `File descriptor 2877 is used by transport` errors.

I've been able to repro with a small script ([gist](https://gist.github.com/top-oai/9c9769e92f65fcc7e4139c170a14cc2f)) that starts/stops/cancels many connections repeatedly. I can repro on at least Ubuntu 24 x86, aiohttp 3.11.11 and uvloop 0.21.0. This fix resolves the errors in my repro script.

I was unable to produce a detailed end-to-end explanation for what's happening here, but I do know that we enter the error state when an exception is thrown while inside of uvloop's `create_connection` function, here:

https://github.com/MagicStack/uvloop/blob/7bb12a174884b2ec8b3162a08564e5fb8a5c6b39/uvloop/loop.pyx#L2066

In that case, if `sock.close()` is not called (this pr changes that), uvloop seems to erroneously hold on to a ref to the transport when the underlying file descriptor has been relinquished to the os.

It feels like at its core a bug in uvloop, however this at least works around it in `aiohttp`.

## Are there changes in behavior for the user?

No... Aside from this error not reproducing.

## Is it a substantial burden for the maintainers to support this?

Unless I've made a mistake (it is a very small change though), I think this is a pure bug fix.

## Related issue number

https://github.com/MagicStack/uvloop/issues/653

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES/` folder
  * name it `<issue_or_pr_num>.<type>.rst` (e.g. `588.bugfix.rst`)
  * if you don't have an issue number, change it to the pull request
    number after creating the PR
    * `.bugfix`: A bug fix for something the maintainers deemed an
      improper undesired behavior that got corrected to match
      pre-agreed expectations.
    * `.feature`: A new behavior, public APIs. That sort of stuff.
    * `.deprecation`: A declaration of future API removals and breaking
      changes in behavior.
    * `.breaking`: When something public is removed in a breaking way.
      Could be deprecated in an earlier release.
    * `.doc`: Notable updates to the documentation structure or build
      process.
    * `.packaging`: Notes for downstreams about unobvious side effects
      and tooling. Changes in the test invocation considerations and
      runtime assumptions.
    * `.contrib`: Stuff that affects the contributor experience. e.g.
      Running tests, building the docs, setting up the development
      environment.
    * `.misc`: Changes that are hard to assign to any of the above
      categories.
  * Make sure to use full sentences with correct case and punctuation,
    for example:
    ```rst
    Fixed issue with non-ascii contents in doctest text files
    -- by :user:`contributor-gh-handle`.
    ```

    Use the past tense or the present tense a non-imperative mood,
    referring to what's changed compared to the last released version
    of this project.

bdraco commented:
It looks like the fix for this in uvloop has stalled out

https://github.com/MagicStack/uvloop/issues/645
https://github.com/MagicStack/uvloop/pull/646

bdraco commented on aiohttp/connector.py:
```diff
@@ -1112,7 +1113,11 @@ async def _wrap_create_connection(
+                connection = await self._loop.create_connection(
+                    *args, **kwargs, sock=sock
+                )
+                sock = None
```
This could use some comments about what it's solving

top-oai replied:
Added a comment, but added it to `sock.close()` since that seemed like a more relevant place for it.

bdraco commented:
Looks like there is a test that will need to be adjusted as well

bdraco commented:
Might also be the same issue reported here https://github.com/MagicStack/uvloop/issues/631

bdraco commented:
I'm really starting to get the feeling uvloop isn't going go to fix this issue anytime soon as there are multiple comments saying they will look at it shortly but it doesn't happen. That's a bit frustrating but it's up to them how they want to prioritize.

bdraco commented on aiohttp/connector.py:
```diff
@@ -1121,6 +1126,9 @@ async def _wrap_create_connection(
                 raise
             raise client_error(req.connection_key, exc) from exc
+        finally:
+            if sock is not None:
```
Explicitly closing the sock is fine, we shouldn't have to do this but it's also fine.

Normally I'd prefer uvloop fix the race. However given how popular uvloop is, how high the impact can be, and the fact a fix is stalled upstream, I think it's reasonable to do the explicit close here.

bdraco commented:
It looks like one of the connector tests is failing with this change.

top-oai commented:
Oh, thanks a lot for all the context...

Yeah this sounds good, I'll circle back to look at the test and add a comment and generally polish this up tomorrow.

top-oai commented:
Noticed the argument about whether it's worth it to maintain this complexity or not, polished up the PR anyway (fixed tests, added new test, added changelog) just so that it's actually ready if we do want it.

> It looks like they are just not around at the moment, there appears to be no activity at all since just after that PR was opened. When they get back to the project, it'll probably be the first thing they do...
I'd much rather uvloop is fixed than need to maintain hacks here.

Yeah that makes sense. A counterargument here is that this generally hardens aiohttp by making it robust against event loop implementations. Also, even with a correct event loop implementation this would enable the socket destruction to potentially be hastened in this case where we know it's garbage.

In any case this PR would unblock our `aiohttp` upgrade, since we are already depending on `uvloop`...  but waiting for `uvloop` to merge the fix on their end certainly works too.

bdraco commented on aiohttp/connector.py:
```diff
@@ -1101,6 +1101,7 @@ async def _wrap_create_connection(
         client_error: Type[Exception] = ClientConnectorError,
         **kwargs: Any,
     ) -> Tuple[asyncio.Transport, ResponseHandler]:
+        sock = None
```
```suggestion
        sock: Union[socket.socket, None] = None
```

A bit nicer to be explicit here

top-oai replied:
Ah, yes definitely. Done.

bdraco commented:
I'm fine with this change given explicitly closing the socket is still a good practice.

bdraco commented:
I'd like to get a 👍 or at a comment from @Dreamsorcerer that he doesn't wan to hold this PR before merging.

bdraco commented on CHANGES/10464.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Changed connection creation to explicitly close sockets if an exception is raised in the event loop's ``create_connection`` method.
```
```suggestion
Changed connection creation to explicitly close sockets if an exception is raised in the event loop's ``create_connection`` method -- by :user:`top-oai`.
```

top-oai replied:
Added.

bdraco commented on tests/test_connector.py:
```diff
@@ -646,6 +646,27 @@ async def test_tcp_connector_certificate_error(
+    req = ClientRequest("GET", URL("https://127.0.0.1:443"), loop=loop)
+
+    conn = aiohttp.TCPConnector()
+    with mock.patch.object(
+        conn._loop,
+        "create_connection",
+        autospec=True,
+        spec_set=True,
+        side_effect=ValueError,
+    ):
+        with pytest.raises(ValueError):
```
with statements can be combined.

Dreamsorcerer replied:
With the first one already over multiple lines, it'll probably be messy combined..

top-oai replied:
I combined them, doesn't seem much much messier than it is currently, but happy to continue to adjust.

Dreamsorcerer commented:
> I'd like to get a 👍 or at a comment from @Dreamsorcerer that he doesn't wan to hold this PR before merging.

I've made one last attempt to ping the uvloop maintainers. If there's no response in a couple of days, then I guess we should merge this.

bdraco commented:
I don't see any activity from uvloop so I'm going to proceed.

Thanks @top-oai

tkukushkin commented:
Hello! Could you please clarify the expected release date for this change? We had to upgrade aiohttp to version 3.10.2+ due to #8597, but we are now encountering this problem, which is significantly impacting our applications.

bdraco commented:
We haven't scheduled a release yet, but I can try to do something this week

bdraco commented:
It looks like this is the cause of the regression in https://github.com/aio-libs/aiohttp/issues/10617 

It looks like we will have to revert this PR

bdraco commented:
The good news is that https://github.com/MagicStack/uvloop/pull/646 does appear to fix this issue as the reproducer no longer works when I manually build uvloop with that fix

PLPeeters commented:
I just ran into this with version 3.12.14 and uvloop 0.21.0; is it possible a regression was introduced somewhere since this PR?

bdraco commented:
This was reverted because it caused problems with asyncio.  https://github.com/MagicStack/uvloop/pull/646 is the fix we are waiting for from upstream
